src/Microsoft.Health.Dicom.Core/Features/Retrieve/RetrieveResourceService.cs

```diff
@@ -63,47 +70,85 @@ public async Task<RetrieveResourceResponse> GetInstanceResourceAsync(RetrieveRes
-                    }));
-
-                long lengthOfResultStreams = resultInstances.Sum(i => i.Stream.Length);
-                _dicomRequestContextAccessor.RequestContext.IsTranscodeRequested = !isOriginalTransferSyntaxRequested;
-                _dicomRequestContextAccessor.RequestContext.BytesTranscoded = isOriginalTransferSyntaxRequested ? 0 : lengthOfResultStreams;
-
-                if (message.ResourceType == ResourceType.Frames)
+                // if single instance, check if we can support transcoding/frame parsing based on fileSize
+                if (retrieveInstances.Count() == 1)
                 {
-                    RetrieveResourceInstance instanceStream = resultInstances.Single();
-                    IReadOnlyCollection<Stream> frames = await _frameHandler.GetFramesResourceAsync(
-                            instanceStream.Stream,
-                            message.Frames,
-                            isOriginalTransferSyntaxRequested,
-                            requestedTransferSyntax);
-
-                    return new RetrieveResourceResponse(
-                        frames.Select(f => new RetrieveResourceInstance(f, instanceStream.TransferSyntaxUid)),
-                        acceptHeaderDescriptor.MediaType);
-                }
-                else if (!isOriginalTransferSyntaxRequested)
-                {
-                    resultInstances = await Task.WhenAll(resultInstances.Select(async x => new RetrieveResourceInstance(await _transcoder.TranscodeFileAsync(x.Stream, requestedTransferSyntax), x.TransferSyntaxUid)));
+                    InstanceMetadata instance = retrieveInstances.First();
+                    var needsTranscoding = NeedsTranscoding(isOriginalTransferSyntaxRequested, requestedTransferSyntax, instance);
+
+                    // need a realized stream if the stream needs to be parsed into a DicomFile
+                    if (needsTranscoding || message.ResourceType == ResourceType.Frames)
+                    {
+                        FileProperties fileProperties = await _blobDataStore.GetFilePropertiesAsync(instance.VersionedInstanceIdentifier, cancellationToken);
+
+                        // limit the file size that can be read in memory
+                        if (fileProperties.ContentLength > _retrieveConfiguration.MaxDicomFileSize)
+                        {
+                            throw new NotAcceptableException(string.Format(DicomCoreResource.RetrieveServiceFileTooBig, _retrieveConfiguration.MaxDicomFileSize));
+                        }
+
+                        // set properties used for billing
+                        if (needsTranscoding)
+                        {
+                            _dicomRequestContextAccessor.RequestContext.IsTranscodeRequested = true;
+                            _dicomRequestContextAccessor.RequestContext.BytesTranscoded = fileProperties.ContentLength;
```

Comment:
Does this mean on a frames request that needs transcoding, we count the entire file as transcoded?

Reply:
good catch, that is a bug, fixed it.